bindings/c/test/mako/future.hpp

```diff
@@ -26,62 +26,103 @@
 #include <string_view>
 #include "logger.hpp"
 #include "macro.hpp"
+#include "stats.hpp"
 
 extern thread_local mako::Logger logr;
 
 namespace mako {
 
-enum class FutureRC { OK, RETRY, CONFLICT, ABORT };
+enum class FutureRC { OK, RETRY, ABORT };
+
+force_inline void updateErrorStats(ThreadStatistics& stats, const fdb::Error& err, int op) {
+	if (err) {
+		if (err.is(1020 /*not_commited*/)) {
+			stats.incrConflictCount();
+		} else {
+			stats.incrErrorCount(op);
+		}
+	}
+}
 
 template <class FutureType>
-force_inline FutureRC handleForOnError(fdb::Transaction& tx, FutureType& f, std::string_view step) {
+force_inline FutureRC handleForOnError(fdb::Transaction& tx,
+                                       FutureType& f,
+                                       std::string_view step,
+                                       ThreadStatistics* stats = nullptr,
+                                       int op = 0) {
 	if (auto err = f.error()) {
-		if (err.is(1020 /*not_committed*/)) {
-			return FutureRC::CONFLICT;
-		} else if (err.retryable()) {
-			logr.warn("Retryable error '{}' found at on_error(), step: {}", err.what(), step);
-			return FutureRC::RETRY;
-		} else {
-			logr.error("Unretryable error '{}' found at on_error(), step: {}", err.what(), step);
-			tx.reset();
-			return FutureRC::ABORT;
+		assert(!(err.retryable()));
+		logr.error("Unretryable error '{}' found at on_error(), step: {}", err.what(), step);
+		tx.reset();
+		if (stats) {
+			updateErrorStats(*stats, err, op);
 		}
+		return FutureRC::ABORT;
 	} else {
 		return FutureRC::RETRY;
 	}
 }
 
 template <class FutureType>
-force_inline FutureRC waitAndHandleForOnError(fdb::Transaction& tx, FutureType& f, std::string_view step) {
+force_inline FutureRC waitAndHandleForOnError(fdb::Transaction& tx,
+                                              FutureType& f,
+                                              std::string_view step,
+                                              ThreadStatistics* stats = nullptr,
+                                              int op = 0,
+                                              bool on_error_future_updates = false) {
 	assert(f);
-	if (auto err = f.blockUntilReady()) {
-		logr.error("'{}' found while waiting for on_error() future, step: {}", err.what(), step);
+	if (!waitFuture(f, step, stats, op)) {
 		return FutureRC::ABORT;
 	}
-	return handleForOnError(tx, f, step);
+
+	return handleForOnError(tx, f, step, on_error_future_updates ? stats : nullptr, op);
 }
 
 // wait on any non-immediate tx-related step to complete. Follow up with on_error().
 template <class FutureType>
-force_inline FutureRC waitAndHandleError(fdb::Transaction& tx, FutureType& f, std::string_view step) {
+force_inline FutureRC waitAndHandleError(fdb::Transaction& tx,
```

Comment:
This is bad.
Default arguments are another way to overload functions and 2 default arguments mean 3 function overloads.
That makes 4 overloads for `waitAndHandleForOnError`.
Not to mention that there seems to be an assumption that `stats` and `op` are a package deal so overloads between maximal and minimal overload and are not used, but adds to the possibility of logic errors.

https://google.github.io/styleguide/cppguide.html#Default_Arguments

Please either make the stats and updates mandatory (and therefore not nullable) or update the stats outside of future utilities.
Let's keep things simple.

Reply:
Another possible solution (if you find overloads truly necessary) is to explicitly overload the function template with non-nullable arguments.